runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1519,7 +1606,368 @@ static iree_status_t iree_hal_hip_device_queue_dealloca(
+      iree_infinite_timeout());
+
+  iree_hal_hip_per_device_info_t* info = &device->devices[device_ordinal];
+  iree_slim_mutex_lock(&info->file_transfer_staging_buffer_lock);
+  out_chunks[0].offset = info->file_transfer_staging_buffer_head;
+  out_chunks[0].size =
+      iree_min(size, device->params.file_transfer_buffer_size -
+                         info->file_transfer_staging_buffer_head);
+  if (size != out_chunks->size) {
+    out_chunks[1].offset = 0;
+    out_chunks[1].size = size - out_chunks[0].size;
+    info->file_transfer_staging_buffer_head = out_chunks[1].size;
+  } else {
+    out_chunks[1].offset = 0;
+    out_chunks[1].size = 0;
+    info->file_transfer_staging_buffer_head += size;
+  }
+  iree_slim_mutex_unlock(&info->file_transfer_staging_buffer_lock);
+}
+
+static iree_status_t iree_hal_hip_device_perform_queue_read_now(
+    void* user_data, iree_status_t status) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_hip_device_semaphore_queue_read_callback_data_t* data =
+      (iree_hal_hip_device_semaphore_queue_read_callback_data_t*)user_data;
+
+  iree_hal_hip_device_t* device = data->base.device;
+
+  // If we had a semaphore failure then we should propagate it
+  // but not run anything.
+  if (!iree_status_is_ok(data->base.status)) {
+    status = iree_status_join(data->base.status, status);
+  }
+
+  int device_ordinal =
+      iree_math_count_trailing_zeros_u64(data->base.queue_affinity);
+
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_CALL_TO_STATUS(
+        device->hip_symbols,
+        hipCtxPushCurrent(device->devices[device_ordinal].hip_context));
+  }
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, device_ordinal);
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_device_stream_wait_for_semaphores(
+        device, data->base.wait_semaphore_list, device_ordinal);
+  }
+  const iree_hal_hip_dynamic_symbols_t* symbols = device->hip_symbols;
+  iree_device_size_t amount_left = data->length;
+  iree_device_size_t offset = 0;
+
+  for (iree_host_size_t i = 0;
+       i < data->num_read_chunks && iree_status_is_ok(status); ++i) {
+    iree_device_size_t chunk_size =
+        iree_min(device->params.file_transfer_chunk_size, amount_left);
+    iree_hal_hip_transfer_buffer_chunk_t chunks[2];
+    iree_hal_hip_transfer_buffer_reserve_chunks(device, device_ordinal,
+                                                chunk_size, &chunks[0]);
+
+    iree_device_size_t read_offset = offset;
+    for (iree_host_size_t j = 0; j < 2; ++j) {
+      if (chunks[j].size) {
+        status = iree_hal_file_read(
+            data->source_file, data->source_offset + read_offset,
+            device->devices[device_ordinal].file_transfer_staging_buffer,
+            chunks[j].offset, chunks[j].size);
+        if (!iree_status_is_ok(status)) {
+          break;
+        }
+        read_offset += chunks[j].size;
+      }
+    }
+
+    // We use a command buffer because it allows us to easily get tracing events
```

Comment:
it pains me a bit to do all of this instead of just scheduling the operation on the stream with a hipMemcpyAsync, but with how the hip HAL and hip itself is designed this is probably fine and still significantly better than the emulated streaming path - future HALs will be designed to avoid the need for this dance, though

Reply:
Yeah, I am not a huge fan (we CAN just use `hipMemcpyAsync`, but I really don't want to lose the tracing information here. Luckily stream command buffers are quite light-weight.